File: The Symmetry Of Gibralfaro Halls/input.cpp
#include <stdio.h>
#include <string.h>

int distance(const char *str1, const char *str2) {
    // Base case: if either string is empty, return the length of the other string
    if (str1[0] == '\0') {
        return strlen(str2);
    }
    if (str2[0] == '\0') {
        return strlen(str1);
    }

    // If the last characters are the same, no edit is needed for this position
    if (str1[0] == str2[0]) {
        return distance(str1 + 1, str2 + 1);
    }

    // Otherwise, consider three possible operations: insertion, deletion, and substitution
    int insertion = distance(str1, str2 + 1);
    int deletion = distance(str1 + 1, str2);
    int substitution = distance(str1 + 1, str2 + 1);

    // Return the minimum of the three operations plus 1
    return 1 + ((insertion < deletion) ? (insertion < substitution ? insertion : substitution)
                                       : (deletion < substitution ? deletion : substitution));
}

int main() {
    char string1[100], string2[100];

    scanf("%s", string1);

    scanf("%s", string2);

    int result = distance(string1, string2);
    printf("distance between '%s' and '%s': %d\n", string1, string2, result);

    return 0;
}

File: The Symmetry Of Gibralfaro Halls/input.pas
Uses math;

function distance(const str1, str2: string): integer;
begin
  // Base case: if either string is empty, return the length of the other string
  if str1 = '' then
    Exit(Length(str2));

  if str2 = '' then
    Exit(Length(str1));

  // If the last characters are the same, no edit is needed for this position
  if str1[Length(str1)] = str2[Length(str2)] then
    Exit(distance(Copy(str1, 1, Length(str1) - 1), Copy(str2, 1, Length(str2) - 1)));

  // Otherwise, consider three possible operations: insertion, deletion, and substitution
  distance := 1 + min(
    distance(str1, Copy(str2, 1, Length(str2) - 1)),        // Insertion
    min(distance(Copy(str1, 1, Length(str1) - 1), str2),       // Deletion
    distance(Copy(str1, 1, Length(str1) - 1), Copy(str2, 1, Length(str2) - 1))) // Substitution
  );
end;

// Example usage
var
  string1, string2: string;
  result: integer;
begin 
  readln(string1);
  readln(string2);
  result := distance(string1, string2);
  writeln('distance between ''', string1, ''' and ''', string2, ''': ', result);
end.